src/rider/main/kotlin/me/fornever/avaloniarider/previewer/AvaloniaPreviewerSessionController.kt

```diff
@@ -124,9 +130,22 @@ class AvaloniaPreviewerSessionController(
 
     @Volatile
     private var session: AvaloniaPreviewerSession? = null
+    @Volatile
+    private var lastKnownProjectRelativePath: String? = null
+    @Volatile
+    private var lastProjectFilePath: Path? = null
+    @Volatile
+    private var pendingRestart = false
 
     private val sessionLifetimeSource = SequentialLifetimes(controllerLifetime)
     private var currentSessionLifetime: LifetimeDefinition? = null
+    private val restartLifetime = controllerLifetime.createNestedDisposable(
+        "AvaloniaPreviewerSessionController.restartLifetime"
+    )
+    private val restartAlarm = Alarm(restartLifetime)
+
+    private val baseDocument: Document? =
+        application.runReadAction<Document?> { FileDocumentManager.getInstance().getDocument(xamlFile) }
```

Comment:
If we need a `Document` that much, let's maybe parametrise the controlled with it (directly pass as a ctor parameter), instead of calculating it in the initializer?